packages/lib/gpcircuits/src/multituple.ts

```diff
@@ -0,0 +1,232 @@
+    lastTuple.push(firstIndex);
+  }
+
+  return tupleIndices;
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'tuple hash',
+ * thus chaining together smaller tuples to form/represent a larger one.
+ * It accomplishes this by hashing the values, then reducing the resulting
+ * tuple down to one element, `paramTupleArity` elements at a time, via
+ * the Poseidon hash function, padding the tuple with the hash of `elements[0]`
+ * if necessary. The precise mechanism is described in `multiTupleHasher` below.
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param elements the values forming the input tuple
+ * @returns appropriately-formed hash of the input tuple
+ * @throws RangeError if `paramTupleArity` is outside the admissible range
+ */
+export function hashTuple(
+  paramTupleArity: number,
+  elements: PODValue[]
+): bigint {
+  // Call the multituple hasher.
+  const multihash = multiTupleHasher(paramTupleArity, elements);
+
+  // Return the last hash.
+  return multihash[multihash.length - 1];
+}
+
+/**
+ * This procedure takes an N-tuple `elements` and computes its 'multituple hash',
+ * which is the result of (Poseidon) hashing the first `paramTupleArity` elements
+ * and subsequently hashing every successive chunk of `paramTupleArity` - 1 elements,
+ * prepending the previously computed hash and padding the remainder of the list
+ * with the first element to form a list of exactly `paramTupleArity` elements
+ * to feed the `tupleHasher`. All of these hashes are returned. This process
+ * mimics the behaviour of the `MultiTupleModule`, which chains together
+ * smaller tuples to form/represent a larger one.
+ *
+ * @param paramTupleArity the arity of the output tuple
+ * @param elements the values forming the input tuple
+ * @returns the list of multituple hashes
+ * @throws RangeError if `paramTupleArity` is outside the admissible range
+ */
+export function multiTupleHasher(
+  paramTupleArity: number,
+  elements: PODValue[]
+): bigint[] {
+  // First hash 'em all.
+  const valueHashes = elements.map(podValueHash);
+
+  // Take note of the first hash, which will be used for padding.
+  const firstValueHash = valueHashes[0];
+
+  // Hash the first `paramTupleArity` elements.
+  const firstChunkHash = tupleHasher(
+    padArray(
+      valueHashes.slice(0, paramTupleArity),
+      paramTupleArity,
+      firstValueHash
+    )
+  );
+
+  // If we have <= `paramTupleArity` elements, return the first
+  // chunk hash.
+  if (valueHashes.length <= paramTupleArity) {
+    return [firstChunkHash];
+  }
+
+  // Else, proceed as follows:
+  const hasher = (tuple: bigint[], hashes: bigint[]): bigint[] =>
+    tuple.length <= paramTupleArity - 1
+      ? // If we have <= `paramTupleArity` - 1 elements, prepend the preceding
+        // chunk hash, pad and hash, appending it to the list of `hashes` computed
+        // so far.
+        hashes.concat(
+          tupleHasher(
```

Comment:
This split of multi-tuple from tuple, mirroring the structure of the cirom templates, helps make this whole process easier to understand.  While analyzing this function I don't need to worry (yet) about the details of how `tupleHasher` does its job.